### PR TITLE
Add 3D depthwise seperable convolution

### PR DESCRIPTION
**Pull request by gungui98**
Because this pull request (#40801) becomes an important part of recent 3D models, brings significant improvement in speed, and also have been open for a while. So I decided to resolve the previous review comment and modify it a bit so that it can be merged into the latest version of Pytorch.

### CODE REVIEW

File: aten/src/ATen/native/Convolution.cpp [hunk 1]
```diff
@@ -684,7 +684,13 @@ at::Tensor _convolution(
             input.contiguous(), weight, bias,
             padding, stride, dilation, params.groups, params.benchmark, params.deterministic);
       } else {
-          output = at::thnn_conv_depthwise2d(input.contiguous(), weight, kernel_size, bias, stride, padding, dilation);
+          if (input.ndimension() == 4) {
```
Comment (albanD) [inline aten/src/ATen/native/Convolution.cpp:687]: Do you actually ever get here if cudnn is available?
Should the cudnn detection be modified to not use it in this particular setting?

**Review by albanD — COMMENTED** [review]
Hi,

Thanks for the updated PR.

I think this is missing documentation, test (both for forward and backward) and some benchmark.

cc @ngimel do you know who would be a good person to review the cuda kernel?

File: aten/src/ATen/native/Convolution.cpp (hunk 1 above)
Comment (gungui98) [reply aten/src/ATen/native/Convolution.cpp:687]: Actually only full precision will go to here, is that okie if add ```input.ndimension() == 4``` condition to ```use_cudnn_depthwise```  function to ensure every 5D tensor input will go to ```conv_depthwise3d```?

**Comment by ngimel** [comment]
Documentation is covered as part of regular convolution documentation (it's not adding new APIs, just efficient implementation for existing op), but we should check if test coverage is sufficient (probably not) and definitely add benchmarks (see documentation for benchmarking utils https://pytorch.org/docs/master/benchmark_utils.html?highlight=timer#torch.utils.benchmark.Timer and and example script in e.g. this PR #50039)

File: aten/src/ATen/native/cuda/DepthwiseConv3d.cu [hunk 2]
```diff
@@ -0,0 +1,634 @@
...
+    if (bias != NULL) {
+      sum += bias[out_channel];
+    }
+
+    output[batch][out_channel][out_frame][out_row][out_col] = sum;
+  }
+}
+
+template <typename scalar_t, typename accscalar_t,
+    int kKnownKernelT, int kKnownKernelH, int kKnownKernelW,
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/DepthwiseConv3d.cu:87]: nit: you are not going to support different templated kernel sizes for different dimensions, so might as well have a single template argument here, same for dilation and stride

File: aten/src/ATen/native/cuda/DepthwiseConv3d.cu [hunk 3]
```diff
@@ -0,0 +1,634 @@
...
+#include <limits>
+
+namespace at {
+namespace native {
+namespace {
+
+template <typename scalar_t, typename accscalar_t,
+    int kKnownKernelT, int kKnownKernelH, int kKnownKernelW,
+    int kKnownDilationT, int kKnownDilationH, int kKnownDilationW>
+__global__ void conv_depthwise3d_cuda_kernel(
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/DepthwiseConv3d.cu:18]: how come there's no stride argument here?

File: aten/src/ATen/native/cuda/DepthwiseConv3d.cu [hunk 4]
```diff
@@ -0,0 +1,634 @@
...
+    const Tensor& weight,
+    IntArrayRef kernel_size,
+    const Tensor& bias,
+    IntArrayRef stride,
+    IntArrayRef padding,
+    IntArrayRef dilation) {
+  TENSOR_ARG(input, 1);
+  TENSOR_ARG(weight, 2);
+
+  checkAllSameGPU("conv_depthwise3d_cuda", {input_arg, weight_arg});
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/DepthwiseConv3d.cu:391]: it looks like you need TensorArgs only to call checkAllSameGPU, but wrapping Tensors in TensorArgs is relatively expensive. You can directly compare devices on Tensors.

**Comment by gungui98** [comment]
Benchmark
```
[---------------------- conv3d mdtype torch.uint8 input dtypetorch.int32 ----------------------]
                            |  After PR DepthWiseConv3d CUDA  |  Before PR DepthWiseConv3d CUDA
1 threads: ------------------------------------------------------------------------------------
      (8, 64, 32, 56, 56)   |               14.8              |              105.5
      (8, 128, 32, 28, 28)  |                7.5              |               52.8
      (8, 256, 32, 14, 14)  |                4.1              |               28.6
      (8, 512, 32, 7, 7)    |                2.7              |               27.7
      (8, 54, 13, 40, 40)   |                2.6              |               32.7
      (8, 108, 13, 20, 20)  |                1.3              |               17.6

Times are in milliseconds (ms).

[--------------------- conv3d mdtype torch.uint8 input dtypetorch.float32 ---------------------]
                            |  After PR DepthWiseConv3d CUDA  |  Before PR DepthWiseConv3d CUDA
1 threads: ------------------------------------------------------------------------------------
      (8, 64, 32, 56, 56)   |               14.9              |              107.8
      (8, 128, 32, 28, 28)  |                7.5              |               53.7
      (8, 256, 32, 14, 14)  |                4.1              |               29.1
      (8, 512, 32, 7, 7)    |                2.7              |               27.9
      (8, 54, 13, 40, 40)   |                2.6              |               33.0
      (8, 108, 13, 20, 20)  |                1.3              |               17.6

Times are in milliseconds (ms).

[--------------------- conv3d mdtype torch.uint8 input dtypetorch.float64 ---------------------]
                            |  After PR DepthWiseConv3d CUDA  |  Before PR DepthWiseConv3d CUDA
1 threads: ------------------------------------------------------------------------------------
      (8, 64, 32, 56, 56)   |               15.0              |              108.9
      (8, 128, 32, 28, 28)  |                7.6              |               54.2
      (8, 256, 32, 14, 14)  |                4.1              |               29.2
      (8, 512, 32, 7, 7)    |                2.7              |               27.9
      (8, 54, 13, 40, 40)   |                2.6              |               33.0
      (8, 108, 13, 20, 20)  |                1.4              |               17.6

Times are in milliseconds (ms).

[----------------------- conv3d mdtype torch.bool input dtypetorch.int32 ----------------------]
                            |  After PR DepthWiseConv3d CUDA  |  Before PR DepthWiseConv3d CUDA
1 threads: ------------------------------------------------------------------------------------
      (8, 64, 32, 56, 56)   |               15.0              |              109.3
      (8, 128, 32, 28, 28)  |                7.6              |               54.4
      (8, 256, 32, 14, 14)  |                4.1              |               29.2
      (8, 512, 32, 7, 7)    |                2.7              |               27.9
      (8, 54, 13, 40, 40)   |                2.6              |               33.0
      (8, 108, 13, 20, 20)  |                1.4              |               17.6

Times are in milliseconds (ms).

[---------------------- conv3d mdtype torch.bool input dtypetorch.float32 ---------------------]
                            |  After PR DepthWiseConv3d CUDA  |  Before PR DepthWiseConv3d CUDA
1 threads: ------------------------------------------------------------------------------------
      (8, 64, 32, 56, 56)   |               15.0              |              109.5
      (8, 128, 32, 28, 28)  |                7.6              |               54.5
      (8, 256, 32, 14, 14)  |                4.1              |               29.3
      (8, 512, 32, 7, 7)    |                2.7              |               28.0
      (8, 54, 13, 40, 40)   |                2.6              |               33.0
      (8, 108, 13, 20, 20)  |                1.4              |               17.6

Times are in milliseconds (ms).

[---------------------- conv3d mdtype torch.bool input dtypetorch.float64 ---------------------]
                            |  After PR DepthWiseConv3d CUDA  |  Before PR DepthWiseConv3d CUDA
1 threads: ------------------------------------------------------------------------------------
      (8, 64, 32, 56, 56)   |               15.0              |              109.5
      (8, 128, 32, 28, 28)  |                7.6              |               54.6
      (8, 256, 32, 14, 14)  |                4.1              |               29.4
      (8, 512, 32, 7, 7)    |                2.7              |               28.1
      (8, 54, 13, 40, 40)   |                2.6              |               33.1
      (8, 108, 13, 20, 20)  |                1.4              |               17.7

Times are in milliseconds (ms). 
```

Benchmark code, modified from example script:
```python
import torch
import itertools
import time

from torch import nn
from torch.utils.benchmark import Timer
from torch.utils.benchmark import Compare
import sys
import pickle

print('Using pytorch %s' % (torch.__version__))

shapes = [(8, 64, 32, 56, 56),
            (8, 128, 32, 28, 28),
            (8, 256, 32, 14, 14),
            (8, 512, 32, 7, 7),
            (8, 54, 13, 40, 40),
            (8, 108, 13, 20, 20),
            ]
results = []
repeats = 10

if __name__ == '__main__':
    for m_dtype in [torch.uint8, torch.bool]:
        for dtype in [torch.int32, torch.float, torch.double]:
            for mat1_shape in shapes:
                mat1 = torch.randn(mat1_shape, requires_grad=True, device="cuda")
                conv = nn.Conv3d(mat1.size(1), mat1.size(1),
                                 kernel_size=(3, 3, 3),
                                 padding=(1, 1, 1),
                                 stride=(1, 1, 1),
                                 groups=mat1.size(1),
                                 bias=False).cuda()
                nn.init.kaiming_normal_(conv.weight)

                tasks = [("conv(mat1).sum().backward()", "After PR DepthWiseConv3d CUDA")]
                timers = [Timer(stmt=stmt, label=f"conv3d mdtype {m_dtype} input dtype{dtype}", sub_label=f"{(mat1_shape)}", description=label, globals=globals()) for stmt, label in tasks]

                for i, timer in enumerate(timers * repeats):
                    results.append(
                        timer.blocked_autorange()
                    )
                    print(f"\r{i + 1} / {len(timers) * repeats}", end="")
                    sys.stdout.flush()

    with open('after_pr.pkl', 'wb') as f:
        pickle.dump(results, f)

    comparison = Compare(results)
    comparison.print()
```

**Comment by gungui98** [comment]
@ngimel ping.

**Comment by ngimel** [comment]
Failing test is real https://app.circleci.com/pipelines/github/pytorch/pytorch/267162/workflows/67e463af-669b-421b-99c9-d1fb234c9194/jobs/10546038, can you please see how double backward is handled for depthwise 2d convolutions, and do similar things here? Also, some tests are failing because of merge conflicts, can you please rebase?
Otherwise looks good, thank you for working on this!

File: aten/src/ATen/native/cuda/DepthwiseConv3d.cu [hunk 5]
```diff
@@ -0,0 +1,624 @@
...
+  }
+}
+
+}
+
+#define NODEF_OR_EQUAL(x, y) ((y) < 0 || (x) == (y))
+#define NODEF_OR_EQUAL_3(x, y1, y2, y3) \
+  (NODEF_OR_EQUAL(x[0], y1) && \
+   NODEF_OR_EQUAL(x[1], y2) && \
+   NODEF_OR_EQUAL(x[1], y3))
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/DepthwiseConv3d.cu:346]: typo here, x[2], please add a test that would catch this

File: aten/src/ATen/native/cuda/DepthwiseConv3d.cu [hunk 6]
```diff
@@ -0,0 +1,624 @@
...
+    IntArrayRef dilation) {
+  TORCH_CHECK(input.device() == weight.device(), "expects input and weight tensors to be on the same device.");
+  if (bias.defined()) {
+    TORCH_CHECK(input.device() == bias.device(), "expects input and bias tensors to be on the same device.");
+  }
+
+  conv_depthwise_shape_check<3>(input, weight, bias, Tensor() /* undefined */,
+                                kernel_size, stride, padding, dilation);
+
+  auto output_size = conv_output_size(input.sizes(), weight.sizes(),
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/DepthwiseConv3d.cu:395]: conv_output_size assumes that 0-the dimension of input is batch dimension, yet you check input.dim() == 4 only later. It looks like 4d input for 3d convolution is no longer supported - the error is thrown in this case.

File: aten/src/ATen/native/cuda/DepthwiseConv3d.cu [hunk 7]
```diff
@@ -0,0 +1,624 @@
...
+    const std::array<bool, 3> output_mask) {
+
+  TORCH_CHECK(grad_output.device() == input.device() && 
+              input.device() == weight.device(), 
+              "expects input, weight and grad_output to be on the same device.");
+  conv_depthwise_shape_check<3>(
+      input, weight, Tensor() /* undefined */, grad_output,
+      kernel_size, stride, padding, dilation);
+
+  bool is_batch = input.dim() == 5;
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/DepthwiseConv3d.cu:524]: same here, I don't think input with less than 5 dims is legal.

File: aten/src/ATen/native/cuda/DepthwiseConv3d.cu [hunk 8]
```diff
@@ -0,0 +1,624 @@
...
+
+  if (k_channel >= kC) {
+    return;
+  }
+
+  const int laneid = threadIdx.x % C10_WARP_SIZE;
+  const int warpid = threadIdx.x / C10_WARP_SIZE;
+  const int nwarps = blockDim.x / C10_WARP_SIZE;
+
+  scalar_t grad = 0;
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/DepthwiseConv3d.cu:214]: should be accscalar_t. Also, it would be better to convert op1 and op2 to accscalar_t before multiplying them to avoid precision loss (similar in other kernels, you are doing multiplication in scalar_t, and that's prone to over/underflow and loss of precision).

**Comment by gungui98** [comment]
@ngimel gentle ping.

**Comment by ngimel** [comment]
ROCM errors 
```
Feb 08 09:47:49 /var/lib/jenkins/workspace/aten/src/ATen/native/hip/DepthwiseConv3d.hip:147:34: error: no matching function for call to '__ldg'
Feb 08 09:47:49             const scalar_t op1 = __ldg(kernel_ptr++);
Feb 08 09:47:49                                  ^~~~~
```
are real. Either don't use `__ldg`, on modern hardware it provides no benefit, you can verify with benchmarks, or guard it with HIP macros and provide non-ldg version for hip.

File: tools/autograd/derivatives.yaml [hunk 9]
```diff
@@ -1541,6 +1541,9 @@
 - name: conv_depthwise3d(Tensor self, Tensor weight, int[3] kernel_size, Tensor? bias, int[3] stride, int[3] padding, int[3] dilation) -> Tensor
   self, weight, bias: conv_depthwise3d_backward(grad.contiguous(), self, weight, kernel_size, stride, padding, dilation, grad_input_mask)
 
+- name: conv_depthwise3d_backward.output_mask(Tensor grad_output, Tensor self, Tensor weight, int[3] kernel_size, int[3] stride, int[3] padding, int[3] dilation, bool[3] output_mask) -> (Tensor grad_input, Tensor grad_weight, Tensor grad_bias)
+  grad_output, self, weight: _convolution_double_backward(grads[0], grads[1], grads[2], grad_output, weight, self, stride, padding, {{1, 1, 1}}, false, {{0, 0, 0}}, 1, false, false, false, false, grad_input_mask)
```
Comment (ngimel) [inline tools/autograd/derivatives.yaml:1545]: why is `groups` argument equal to `1` here? It's self.size(1) for 2d conv, that's what it should be, right?

**Comment by gungui98** [comment]
@ngimel it looks like some tests are failed because of modification in upstream, do I need to rebase again?

**Comment by ngimel** [comment]
Yes, please rebase. Thank you! We should be able to merge soon.

**Comment by gungui98** [comment]
@ngimel the test looks good. What should I do with coverage

**Comment by ngimel** [comment]
Looks like there are no depthwise 3d tests, can you add a test similar to https://github.com/pytorch/pytorch/blob/c4a8f0ceaa24f3922b1e0d96c75bb9c2af9c7e4a/test/test_nn.py#L4617, or modify that one to also test 3d

**Comment by gungui98** [comment]
@ngimel I just add a test for conv3d by simply extend ```test_Conv2d_depthwise_naive_groups_cuda``` to 3 dimensions, but it still not pass the coverage test.

**Comment by ngimel** [comment]
It's ok then, we can ignore coverage failure.

**Review by ngimel — APPROVED** [review]

**Comment by nik123** [comment]
Any chance it will become a part of PyTorch 1.8 release?

**Comment by ngimel** [comment]
Sorry, it won't be part of 1.8 release.

**Comment by dragen1860** [comment]
Will it become a part of PyTorch 1.9 release, so that we do not need to install building from source? thank you.

**Comment by WUSHUANGPPP** [comment]
Good practice！As the comments mentioned, will it become a part of PyTorch1.9(stable) ？@ngimel Thank you！

**Comment by WUSHUANGPPP** [comment]
image
ok，got it~ It was mentioned in the release note of PyTorch 1.9

**Comment by gungui98** [comment]
I made a separate python package so anyone can build with older PT and cudnn versions. https://github.com/gungui98/Depthwise-Conv3d

**Comment by makecent** [comment]
@gungui98 I am using pytorch 1.11.0 and the group convolution still run very slow. For example, with group convolution one can reduce the FLOPs of a model 20x fewer, but the inference runtime is only reduced about a half.
